man/man1/gfstat.1
'\" t
.\"     Title: gfstat
.\"    Author: [FIXME: author] [see http://docbook.sf.net/el/author]
.\" Generator: DocBook XSL Stylesheets v1.76.1 <http://docbook.sf.net/>
.\"      Date: 12 Nov 2015
.\"    Manual: Gfarm
.\"    Source: Gfarm
.\"  Language: English
.\"
.TH "GFSTAT" "1" "12 Nov 2015" "Gfarm" "Gfarm"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
gfstat \- display file or directory status
.SH "SYNOPSIS"
.HP \w'\fBgfstat\fR\ 'u
\fBgfstat\fR [\fIoptions\fR] \fIpath\fR...
.SH "DESCRIPTION"
.PP
\fBgfstat\fR
displays information about the specified files or directories in the Gfarm filesystem\&.
.SH "OPTIONS"
.PP
\fB\-M\fR
.RS 4
displays metadata server information as well\&.
.RE
.PP
\fB\-c\fR
.RS 4
displays only the number of file replicas\&.
.RE
.PP
\fB\-l\fR
.RS 4
does not follow symbolic links, but show the link information itself\&.
.RE
.PP
\fB\-r\fR
.RS 4
shows file size and time of the file data instead of the metadata\&. This option is useful to know the file size during the file creation\&.
.RE
.PP
\fB\-h\fR \fIhost\fR
.RS 4
specifies the location of the file replica\&. This option is only effective with the \-r option\&.
.RE
.PP
\fB\-?\fR
.RS 4
displays a list of command options\&.
.RE


ChangeLog
2007-03-31  Noriyuki Soda  <soda@sra.co.jp>

	* RELNOTES
	avoid an expression which may be misinterpreted as a security problem.

2007-03-31  Osamu Tatebe <tatebe@cs.tsukuba.ac.jp>

	* RELNOTES
	correct description

2007-03-30  Tohru Sotoyama <sotoyama@sra.co.jp>

	* package/solaris/build
	change globus root directory

2007-03-30  Noriyuki Soda  <soda@sra.co.jp>

	* ChangeLog LICENSE RELNOTES
	merge changes in the CVS main trunk until gfarm-1.4.1 to the GNEO
	branch
	
	* ChangeLog
	Catched up with the current ChangeLog.

	* ChangeLog
	recover missed log messages in previous commit
	
	* doc/html/en/ref/man5/gfarm.conf.5.html, doc/html/ja/ref/man5/gfarm.conf.5.html man/ja/man5/gfarm.conf.5 man/man5/gfarm.conf.5:
	regen

	* doc/Gfarm-FAQ.en, doc/Gfarm-FAQ.ja, doc/docbook/en/ref/man5/gfarm.conf.5.docbook, doc/docbook/ja/ref/man5/gfarm.conf.5.docbook:
	merge the following changes in the SourceForge Subversion repository:

	Revision: 3674
		  http://gfarm.svn.sourceforge.net/gfarm/?rev=3674&view=rev
	Author:   sotoyama
	Date:     2007-03-29 02:34:39 -0700 (Thu, 29 Mar 2007)
	Log Message:
	add limitation about mixing gsi authentication with sharedsecret
	found in [gfarm-developers:02814]

	Revision: 3676
		  http://gfarm.svn.sourceforge.net/gfarm/?rev=3676&view=rev
	Author:   sotoyama
	Date:     2007-03-29 04:46:11 -0700 (Thu, 29 Mar 2007)
	Log Message:
	add limitation about mixing gsi authentication with sharedsecret
	found in [gfarm-developers:02814] to `auth' statement

	Revision: 3682
		  http://gfarm.svn.sourceforge.net/gfarm/?rev=3682&view=rev
	Author:   n-soda
	Date:     2007-03-29 11:03:31 -0700 (Thu, 29 Mar 2007)
	Log Message:
	swap section order of limitation and example of the "auth" directive,
	and add a paragraph to describe the example.

2007-03-30  yusuke Tanimura <yusuke.tanimura@aist.go.jp>

	* RELNOTES:
	proofreading

2007-03-29  Noriyuki Soda  <soda@sra.co.jp>

	* server/gfsd/gfsd.c:
	merge the follownig change from the SourceForge Subversion repository:

	Revision: 3681
		  http://gfarm.svn.sourceforge.net/gfarm/?rev=3681&view=rev
	Author:   n-soda
	Date:     2007-03-29 07:37:36 -0700 (Thu, 29 Mar 2007)
	Log Message:
	report error status of close(2) in gfs_server_open_local(),
	problem reported by Markus Elfring as:
	http://sourceforge.net/tracker/index.php?func=detail&aid=1687850&group_id=172718&atid=862742

	Also, treat an error at sendmsg(2) as just an error instead of
	fatal one.

	* bench/fsysbench/fsysbench.c:
	merge the follownig change from the SourceForge Subversion repository:

	Revision: 3679
		  http://gfarm.svn.sourceforge.net/gfarm/?rev=3679&view=rev
	Author:   n-soda
	Date:     2007-03-29 05:13:05 -0700 (Thu, 29 Mar 2007)
	Log Message:
	timer_interrupt_handler() shouldn't call printf(3), but write(2),
	because the former is not async-signal-safe.
	c.f.
	https://sourceforge.net/tracker/?func=detail&atid=862742&aid=1687842&group_id=172718
	
	* ftool/gfarm_agent/gfarm_agent.c:
	merge the follownig change from the SourceForge Subversion repository:

	Revision: 3678
		  http://gfarm.svn.sourceforge.net/gfarm/?rev=3678&view=rev
	Author:   n-soda
	Date:     2007-03-29 04:59:54 -0700 (Thu, 29 Mar 2007)
	Log Message:
	sigterm_handler() shouldn't call exit(3), but _exit(2), because
	the former is not async-signal-safe.
	c.f.
	https://sourceforge.net/tracker/?func=detail&atid=862742&aid=1687842&group_id=172718
	
	* server/gfsd/gfsd.c:
	merge the follownig change from the SourceForge Subversion repository:

	Revision: 3675
		  http://gfarm.svn.sourceforge.net/gfarm/?rev=3675&view=rev
	Author:   n-soda
	Date:     2007-03-29 03:30:02 -0700 (Thu, 29 Mar 2007)
	Log Message:
	do not call exit(3) but _exit(2) in accepting_sigterm_handler(),
	because the former is not async-signal-safe.
	problem reported by Markus Elfring in
	http://sourceforge.net/tracker/index.php?func=detail&aid=1687842&group_id=172718&atid=862742

2007-03-29  yusuke Tanimura <yusuke.tanimura@aist.go.jp>

	* RELNOTES, LICENSE:
	update release notes and the copyright display for the version 1.4.1

2007-03-28  Osamu Tatebe <tatebe@cs.tsukuba.ac.jp>

	* ChangeLog:
	Catched up with the current ChangeLog.

2007-03-28  Noriyuki Soda  <soda@sra.co.jp>

	* include/gfarm/gfarm_misc.h, package/freebsd/Makefile,	package/pkgsrc/Makefile, package/redhat/gfarm.spec, package/solaris/build:
	increase version number to 1.4.1

	* include/gfarm/gfarm_error.h, lib/libgfarm/gfarm/error.c, lib/libgfarm/gfarm/gfs_client.c:
	merge the following change from the SourceForge Subversion repository:

	Revision: 3673
		  http://gfarm.svn.sourceforge.net/gfarm/?rev=3673&view=rev
	Author:   n-soda
	Date:     2007-03-28 07:11:37 -0700 (Wed, 28 Mar 2007)
	Log Message:
	make gfsd reconnection feature work against more network errors.
	this fixes a problem reported by Yamamoto Naotaka-san
	in [gfarm-developers:03456] which was caused by ECONNRESET.

	introduce ENETDOWN, EHOSTDOWN and ENETRESET as gfarm native errors
	for this purpose. (although ENETRESET will be nearly never used)

	* gftool/gneo/gneo.c:
	just a portability fix

	* regress/gfs_hook/lib/fsystest/index.FreeBSD.expected, regress/gfs_hook/lib/fsystest/index.Linux.expected, regress/gfs_hook/lib/fsystest/index.NetBSD.expected, regress/gfs_hook/lib/fsystest/index.SunOS.expected:
	results are different on the GNEO branch with gfs_hook


gftool/gfstat/gfstat.c
/*
 * $Id$
 */

#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <libgen.h>

#include <gfarm/gfarm.h>

#include "gfm_client.h"
#include "lookup.h"
#include "gfarm_path.h"
#include "gfs_pio.h"

/*
 * Display struct gfs_stat.
 */

void
display_gfm_server(struct gfm_connection *gfm_server)
{
	printf("MetadataHost: %s\n", gfm_client_hostname(gfm_server));
	printf("MetadataPort: %d\n", gfm_client_port(gfm_server));
	printf("MetadataUser: %s\n", gfm_client_username(gfm_server));
}

static void
display_time(const char *name, struct gfarm_timespec *ts)
{
#if 1	/* support nanosecond */
	char s[64];
	time_t t = ts->tv_sec;
	struct tm *tm = localtime(&t);

	strftime(s, sizeof(s), "%Y-%m-%d %H:%M:%S", tm);
	printf("%s: %s.%09d", name, s, ts->tv_nsec);
	strftime(s, sizeof(s), "%z", tm);
	printf(" %s\n", s);
#else	/* old format */
	time_t t = ts->tv_sec;

	printf("%s: %s", name, ctime(&t));
#endif
}

void
display_stat(char *fn, struct gfs_stat *st)
{
	printf("  File: \"%s\"\n", fn);
	printf("  Size: %-12" GFARM_PRId64 " Filetype: ", st->st_size);
	switch (st->st_mode & GFARM_S_IFMT) {
	case GFARM_S_IFREG:
		puts("regular file");
		break;
	case GFARM_S_IFDIR:
		puts("directory");
		break;
	case GFARM_S_IFLNK:
		puts("symbolic link");
		break;
	default:
		printf("unknown\n");
	}
	printf("  Mode: (%04o) Uid: (%8s) Gid: (%8s)\n",
	       st->st_mode & GFARM_S_ALLPERM,
	       st->st_user, st->st_group);
	printf(" Inode: %-12" GFARM_PRId64 " Gen: %-12" GFARM_PRId64 "\n",
	       st->st_ino, st->st_gen);
	printf("       (%016llX%016llX)\n",
	       (long long)st->st_ino, (long long)st->st_gen);
	printf(" Links: %-12" GFARM_PRId64 " Ncopy: %-12" GFARM_PRId64 "\n",
	       st->st_nlink, st->st_ncopy);

	display_time("Access", &st->st_atimespec);
	display_time("Modify", &st->st_mtimespec);
	display_time("Change", &st->st_ctimespec);
}

void
display_ncopy(char *fn, struct gfs_stat *st)
{
	printf("%" GFARM_PRId64 "\n", st->st_ncopy);
}

gfarm_error_t
gfs_stat_realsize(char *path, char *host, struct gfs_stat *st)
{
	GFS_File gf;
	gfarm_error_t e;

	e = gfs_pio_open(path, GFARM_FILE_RDONLY, &gf);
	if (e != GFARM_ERR_NO_ERROR)
		return (e);
	/* XXX FIXME: INTERNAL FUNCTION SHOULD NOT BE USED */
	if (host == NULL || (e = gfs_pio_internal_set_view_section(gf, host))
	    == GFARM_ERR_NO_ERROR)
		e = gfs_pio_stat(gf, st);
	(void)gfs_pio_close(gf);
	return (e);
}

void
usage(char *prog_name)
{
	fprintf(stderr, "Usage: %s [option] file1 [file2 ...]\n",
		prog_name);
	fprintf(stderr, "option:\n");
	fprintf(stderr, "\t-M\t\tdisplay metadata server information too\n");
	fprintf(stderr, "\t-c\t\tdisplay number of file replicas only\n");
	fprintf(stderr, "\t-l\t\tshow symbolic links\n");
	fprintf(stderr, "\t-r\t\tshow file size of file data\n");
	fprintf(stderr, "\t-h host\t\tspecify the file system node\n");
	exit(2);
}

/*
 *
 */

int
main(int argc, char *argv[])
{
	char *prog_name = argc > 0 ? basename(argv[0]) : "gfstat";
	char *host = NULL;
	gfarm_error_t e;
	int show_gfm_server = 0, show_ncopy_only = 0, show_symlink = 0;
	int show_realsize = 0;
	int c, first_entry = 1, r = 0;

	while ((c = getopt(argc, argv, "Mch:lr?")) != -1) {
		switch (c) {
		case 'M':
			show_gfm_server = 1;
			break;
		case 'c':
			show_ncopy_only = 1;
			break;
		case 'h':
			host = optarg;
			break;
		case 'l':
			show_symlink = 1;
			break;
		case 'r':
			show_realsize = 1;
			break;
		case '?':
		default:
			usage(prog_name);
		}
	}
	argc -= optind;
	argv += optind;

	if (argc < 1)
		usage(prog_name);
	if (show_gfm_server && show_ncopy_only) {
		fprintf(stderr, "%s: cannot specify -M and -c at once",
		    prog_name);
		usage(prog_name);
	}

	e = gfarm_initialize(&argc, &argv);
	if (e != GFARM_ERR_NO_ERROR) {
		fprintf(stderr, "%s: %s\n", prog_name, gfarm_error_string(e));
		exit(1);
	}

	for (; *argv; ++argv) {
		struct gfs_stat st;
		struct gfm_connection *gfm_server;
		char *realpath = NULL;

		e = gfarm_realpath_by_gfarm2fs(*argv, &realpath);
		if (e == GFARM_ERR_NO_ERROR)
			*argv = realpath;
		if (show_symlink)
			e = gfs_lstat(*argv, &st);
		else if (show_realsize)
			e = gfs_stat_realsize(*argv, host, &st);
		else
			e = gfs_stat(*argv, &st);
		if (e != GFARM_ERR_NO_ERROR) {
			fprintf(stderr, "%s: %s\n", *argv,
				gfarm_error_string(e));
			free(realpath);
			r = 1;
			continue;
		}

		gfm_server = NULL;
		if (show_gfm_server && (e = (*
		    (show_symlink
		    ? gfm_client_connection_and_process_acquire_by_path
		    : gfm_client_connection_and_process_acquire_by_path_follow)
		    )(*argv, &gfm_server)) != GFARM_ERR_NO_ERROR) {
			fprintf(stderr, "%s: showing metadata server: %s",
			    *argv, gfarm_error_string(e));
			free(realpath);
			r = 1;
			continue;
		}

		if (show_ncopy_only) {
			display_ncopy(*argv, &st);
		} else {
			if (!first_entry)
				putchar('\n');
			display_stat(*argv, &st);
		}
		if (gfm_server != NULL) {
			display_gfm_server(gfm_server);
			gfm_client_connection_free(gfm_server);
		}

		gfs_stat_free(&st);
		free(realpath);
		first_entry = 0;
	}

	e = gfarm_terminate();
	if (e != GFARM_ERR_NO_ERROR) {
		fprintf(stderr, "%s: %s\n", prog_name, gfarm_error_string(e));
		exit(1);
	}

	exit(r);
}
